File: python/lcs.py
def lcs_length(X, Y):
    m = len(X)
    n = len(Y)
    c = [[0] * (n + 1) for _ in range(m + 1)]
    b = [[" "] * (n + 1) for _ in range(m + 1)]
    
    for i in range(m):
        for j in range(n):
            if X[i] == Y[j]:
                c[i + 1][j + 1] = c[i][j] + 1
                b[i + 1][j + 1] = "↖"
            elif c[i + 1][j] >= c[i][j + 1]:
                c[i + 1][j + 1] = c[i + 1][j]
                b[i + 1][j + 1] = "←"  
            else:
                c[i + 1][j + 1] = c[i][j + 1]
                b[i + 1][j + 1] = "↑" 
    return c, b

def print_lcs(X, b, i, j):
    if i == 0 or j == 0:
        return
    if b[i][j] == "↖":
        print_lcs(X, b, i - 1, j - 1)
        print(X[i - 1], end="")
    elif b[i][j] == "←":
        print_lcs(X, b, i, j - 1)    
    elif b[i][j] == "↑":
        print_lcs(X, b, i - 1, j)

if __name__ == "__main__":
    X = input("Enter the first string: ")
    Y = input("Enter the second string: ")
    c, b = lcs_length(X, Y)

    print("\nDP Table (c):")
    for row in c:
        print(row)
    
    print("\nDirection Table (b):")
    for row in b:
        print(row)
    
    print("\nLength of LCS:", c[-1][-1])
    print("LCS: ", end="")
    print_lcs(X, b, len(X), len(Y))

                





    

File: cpp/LCS.cpp
#include <iostream>
#include <string>
using namespace std;

#define MAX_LEN 100  

void printLCS(string X, char b[][MAX_LEN], int i, int j) {
    if (i == 0 || j == 0) {
        return;
    }
    if (b[i][j] == 'D') {  // Diagonal arrow (↖)
        printLCS(X, b, i - 1, j - 1);
        cout << X[i-1];
    } else if (b[i][j] == 'L') {  // Left arrow (←)
        printLCS(X, b, i, j - 1);
    } else {  // Up arrow (↑)
        printLCS(X, b, i - 1, j);
    }
}

int main() {
    string X, Y;
    cout << "Enter the first string: ";
    cin >> X;
    cout << "Enter the second string: ";
    cin >> Y;

    int m = X.length();
    int n = Y.length();
    
    int c[MAX_LEN][MAX_LEN] = {0};
    char b[MAX_LEN][MAX_LEN] = {' '};

    for (int i = 0; i < m; i++) {
        for (int j = 0; j < n; j++) {
            if (X[i] == Y[j]) {
                c[i+1][j+1] = c[i][j] + 1;
                b[i+1][j+1] = 'D'; 
            } else if (c[i+1][j] >= c[i][j+1]) {
                c[i+1][j+1] = c[i+1][j];
                b[i+1][j+1] = 'L';
            } else {
                c[i+1][j+1] = c[i][j+1];
                b[i+1][j+1] = 'U';  
            }
        }
    }

    
    cout << "\nDP Table (c):\n";
    for (int i = 0; i <= m; i++) {
        for (int j = 0; j <= n; j++) {
            cout << c[i][j] << " ";
        }
        cout << endl;
    }

    
    cout << "\nDirection Table (b):\n";
    for (int i = 0; i <= m; i++) {
        for (int j = 0; j <= n; j++) {
            cout << b[i][j] << " ";
        }
        cout << endl;
    }

    cout << "\nLength of LCS: " << c[m][n] << endl;
    cout << "LCS: ";
    printLCS(X, b, m, n);
    cout << endl;

    return 0;
}
